src/store_client.cc

```diff
@@ -499,58 +517,126 @@ store_client::fileRead()
     assert(!flags.disk_io_pending);
     flags.disk_io_pending = true;
 
+    // mem->swap_hdr_sz is zero here during initial read(s)
+    const auto nextStoreReadOffset = NaturalSum<int64_t>(mem->swap_hdr_sz, nextHttpReadOffset()).value();
+
+    // XXX: If fileRead() is called when we do not yet know mem->swap_hdr_sz,
+    // then we must read from disk offset zero to learn it. We cannot compute
+    // correct HTTP response start offset on disk without it. However, late
+    // startSwapin() calls imply that the assertion below may fail.
+    Assure(mem->swap_hdr_sz > 0 || !nextStoreReadOffset);
+
+    // TODO: Remove this assertion. Introduced in 1998 commit 3157c72, it
+    // assumes that swapped out memory is freed unconditionally, but we no
+    // longer do that because trimMemory() path checks lowestMemReaderOffset().
+    // It is also misplaced: We are not swapping out anything here and should
+    // not care about any swapout invariants.
     if (mem->swap_hdr_sz != 0)
         if (entry->swappingOut())
-            assert(mem->swapout.sio->offset() > copyInto.offset + (int64_t)mem->swap_hdr_sz);
+            assert(mem->swapout.sio->offset() > nextStoreReadOffset);
+
+    // XXX: We should let individual cache_dirs limit the read size instead, but
+    // we cannot do that without more fixes and research because:
+    // * larger reads crash code that uses SharedMemory::get() (TODO: confirm!)
+    // * we do not know how to find all I/O code that assumes this limit
+    // * performance effects of larger disk reads may be negative somewhere.
+    const decltype(StoreIOBuffer::length) maxReadSize = SM_PAGE_SIZE;
```

Comment:
Please see if you can find a way to crash SharedMemory-using code by artificially requesting a larger I/O size here (e.g., increasing HTTP_REQBUF_SZ to, say, 64KB and removing this limit/overwrite). This task has a high priority.

Reply:
Yes, this is possible. SharedMemory::get() returns an SHMBUF_BLKSZ-miltiple offset (where SHMBUF_BLKSZ=4096). I performed a  manual test, caching 2 entries in diskd and then accessing them simultaneously (I used response delay pools to slow things down). Squid did not crash, both responses were delivered, but one of the responses contained incorrect data.